src/AsyncMethodTester.mo

```diff
@@ -0,0 +1,171 @@
+import Debug "mo:base/Debug";
+import Error "mo:base/Error";
+import Option "mo:base/Option";
+import Buffer "mo:base/Buffer";
+
+module {
+  public type State = {
+    #staged;
+    #running;
+    #ready;
+  };
+
+  public type ReleaseState = {
+    release : () -> ();
+    state : () -> State;
+  };
+
+  class Response<T, S, R>(method : ??(T -> S, S -> R), limit : Nat) {
+    var lock = true;
+    public var state : State = #staged;
+    public var methods : {
+      #none;
+      #error;
+      #some : (T -> S, S -> R);
+    } = switch (method) {
+      case (??x) #some(x);
+      case (?null) #error;
+      case (null) #none;
+    };
+    public var result : ?R = null;
+
+    public func release() {
+      if (not lock) {
+        Debug.trap("Response must be locked before release");
+      };
+      lock := false;
+    };
+
+    public func run(arg : T) : async () {
+      let s = switch (methods) {
+        case (#some(pre, _)) ?pre(arg);
+        case (_) null;
+      };
+
+      var inc = limit;
+      state := #running;
+      while (lock and inc > 0) {
+        await async ();
+        inc -= 1;
+      };
+      state := #ready;
+      if (inc == 0) {
+        Debug.trap("Iteration limit reached");
+      };
+
+      switch (methods, s) {
+        case (#some(_, after), ?state) {
+          result := ?after(state);
```

Comment:
Could you use a different name than `state` because `state` is already used above for the `State` type (staged/running/ready). This is a different state.